src/meta/raft-store/src/state_machine/sm.rs

```diff
@@ -809,43 +900,109 @@ impl StateMachine {
         Ok(new_value.0)
     }
 
+    /// Execute an upsert-kv operation on a transactional sled tree.
+    ///
+    /// KV has two indexes:
+    /// - The primary index: `key -> (seq, meta(expire_time), value)`,
+    /// - and a secondary expiration index: `(expire_time, seq) -> key`.
+    ///
+    /// Thus upsert a kv record is done in two steps:
+    /// update the primary index and optionally update the secondary index.
+    ///
+    /// It returns 3 SeqV:
+    /// - `(None, None, x)`: upsert nonexistent key;
+    /// - `(None, Some, x)`: upsert existent and non-expired key;
+    /// - `(Some, None, x)`: upsert existent but expired key;
     #[allow(clippy::type_complexity)]
     fn txn_upsert_kv(
-        &self,
         txn_tree: &TransactionSledTree,
         upsert_kv: &UpsertKV,
         log_time_ms: u64,
-    ) -> Result<(Option<SeqV>, Option<SeqV>), MetaStorageError> {
+    ) -> Result<(Option<SeqV>, Option<SeqV>, Option<SeqV>), MetaStorageError> {
+        let (expired, prev, res) =
+            Self::txn_upsert_kv_primary_index(txn_tree, upsert_kv, log_time_ms)?;
+
+        let expires = txn_tree.key_space::<Expire>();
+
+        if let Some(sv) = &expired {
+            if let Some(m) = &sv.meta {
+                if let Some(exp) = m.expire_at {
+                    expires.remove(&ExpireKey::new(exp * 1000, sv.seq))?;
+                }
+            }
+        }
+
+        // No change, no need to update expiration index
+        if prev == res {
+            return Ok((expired, prev, res));
+        }
+
+        // Remove previous expiration index, add a new one.
+
+        if let Some(sv) = &prev {
```

Comment:
Let's play with `#![feature(let_chains)]`!

```
if let Some(sv) = &prev && let Some(m) = &sv.meta && let Some(exp) = m.expire_at {
    expires.remove(&ExpireKey::new(exp * 1000, sv.seq))?;
}
```

Reply:
Smart guy!

Reply:
Hmm... Clion does not recognize such syntax and refuses to format codes for me.

I'd have to leave it as is.

Reply:
Tracked at https://github.com/intellij-rust/intellij-rust/issues/8625

Reply:
<img width="108" alt="image" src="https://user-images.githubusercontent.com/44069/198964785-163ded3e-1641-45b1-acc7-2f9e42b4f40d.png">